src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReader.cs

```diff
@@ -1117,6 +1117,60 @@ public AssemblyReference GetAssemblyReference(AssemblyReferenceHandle handle)
             return new AssemblyReference(this, handle.Value);
         }
 
+        internal AssemblyName GetAssemblyName()
+        {
+            StringHandle nameHandle = AssemblyTable.GetName();
+
+            string name = (!nameHandle.IsNil) ? GetString(nameHandle) : null;
+
+            Version version = AssemblyTable.GetVersion();
+
+            StringHandle cultureHandle = AssemblyTable.GetCulture();
+            string cultureName = (!cultureHandle.IsNil) ? GetString(cultureHandle) : null;
+
+            BlobHandle publicKeyHandle = AssemblyTable.GetPublicKey();
+            byte[] publicKeyOrToken = !publicKeyHandle.IsNil ? GetBlobBytes(publicKeyHandle) : null;
+
+            AssemblyHashAlgorithm assemblyHashAlgorithm = AssemblyTable.GetHashAlgorithm();
+
+            AssemblyFlags flags = AssemblyTable.GetFlags();
+
+            var assemblyName = new AssemblyName(name)
+            {
+                Version = version,
+                CultureName = cultureName,
+                HashAlgorithm = (Configuration.Assemblies.AssemblyHashAlgorithm)assemblyHashAlgorithm
+            };
+            assemblyName.SetPublicKey(publicKeyOrToken);
+            SetFlags(assemblyName, flags);
+
+            return assemblyName;
+        }
+
+        private void SetFlags(AssemblyName assemblyName, AssemblyFlags flags)
+        {
+            if ((flags & AssemblyFlags.PublicKey) != 0)
+                assemblyName.Flags |= AssemblyNameFlags.PublicKey;
+            else
+                assemblyName.Flags &= AssemblyNameFlags.PublicKey;
+
+            if ((flags & AssemblyFlags.Retargetable) != 0)
+                assemblyName.Flags |= AssemblyNameFlags.Retargetable;
+            else
+                assemblyName.Flags &= AssemblyNameFlags.Retargetable;
+
+            if ((flags & AssemblyFlags.EnableJitCompileTracking) != 0)
+                assemblyName.Flags |= AssemblyNameFlags.EnableJITcompileTracking;
+            else
+                assemblyName.Flags &= AssemblyNameFlags.EnableJITcompileTracking;
+
+            // notice we are setting EnableJITcompileOptimizer from DisableJitCompileOptimizer. hasFlag Logic is flipped
+            if ((flags & AssemblyFlags.DisableJitCompileOptimizer) == 0)
+                assemblyName.Flags |= AssemblyNameFlags.EnableJITcompileOptimizer;
+            else
+                assemblyName.Flags &= AssemblyNameFlags.EnableJITcompileOptimizer;
```

Comment:
The reason for the flip here is that AssemblyNameFlags has a hilarious naming bug. 

See https://msdn.microsoft.com/en-us/library/system.reflection.assemblynameflags(v=vs.110).aspx (Emphasis mine.):

> EnableJITcompileOptimizer
>
>Specifies that just-in-time (JIT) compiler optimization is disabled for the assembly. *This is the exact opposite of the meaning that is suggested by the member name.*

So, in fact, the true meaning of the bit is the same between both enumerations. Furthermore, both use the numeric value specified by ECMA 335 and you can therefore just cast here too.